internal/batch/batch_processor.go

```diff
@@ -445,81 +476,106 @@ func (bp *batchProcessor) maskContexts(ctx context.Context, batch *fftypes.Batch
 				// Generate a new Transaction, which will be used to record status of the associated transaction as it happens
-				if contexts, err = bp.maskContexts(ctx, batch); err != nil {
+				if state.Pins, err = bp.maskContexts(ctx, &state.Payload); err != nil {
 					return err
 				}
 			}
 
-			batch.Payload.TX.Type = bp.conf.txType
-			if batch.Payload.TX.ID, err = bp.txHelper.SubmitNewTransaction(ctx, batch.Namespace, bp.conf.txType); err != nil {
+			state.Persisted.TX.Type = bp.conf.txType
+			if state.Persisted.TX.ID, err = bp.txHelper.SubmitNewTransaction(ctx, state.Persisted.Namespace, bp.conf.txType); err != nil {
 				return err
 			}
+			state.Payload.TX = state.Persisted.TX
+			state.Manifest.TX = state.Persisted.TX
+
+			// The hash of the batch, is the hash of the manifest to minimize the compute cost.
+			// Note in v0.13 and before, it was the hash of the payload - so the inbound route has a fallback to accepting the full payload hash
+			manifestString := state.Manifest.String()
+			state.Persisted.Manifest = fftypes.JSONAnyPtr(manifestString)
+			state.Persisted.Hash = fftypes.HashString(manifestString)
+
+			log.L(ctx).Debugf("Batch %s sealed. Hash=%s", state.Persisted.ID, state.Persisted.Hash)
 
-			batch.Hash = batch.Payload.Hash()
-			log.L(ctx).Debugf("Batch %s sealed. Hash=%s", batch.ID, batch.Hash)
-			return bp.database.UpsertBatch(ctx, batch)
+			// At this point the manifest of the batch is finalized. We write it to the database
+			return bp.database.UpsertBatch(ctx, &state.Persisted)
 		})
 	})
-	return contexts, err
+	return err
 }
 
-func (bp *batchProcessor) dispatchBatch(batch *fftypes.Batch, pins []*fftypes.Bytes32) error {
+func (bp *batchProcessor) dispatchBatch(state *DispatchState) error {
 	// Call the dispatcher to do the heavy lifting - will only exit if we're closed
 	return operations.RunWithOperationCache(bp.ctx, func(ctx context.Context) error {
 		return bp.retry.Do(ctx, "batch dispatch", func(attempt int) (retry bool, err error) {
-			return true, bp.conf.dispatch(ctx, batch, pins)
+			return true, bp.conf.dispatch(ctx, state)
 		})
 	})
 }
 
-func (bp *batchProcessor) markMessagesDispatched(batch *fftypes.Batch) error {
+func (bp *batchProcessor) markPayloadDispatched(state *DispatchState) error {
 	return bp.retry.Do(bp.ctx, "mark dispatched messages", func(attempt int) (retry bool, err error) {
 		return true, bp.database.RunAsGroup(bp.ctx, func(ctx context.Context) (err error) {
 			// Update all the messages in the batch with the batch ID
-			msgIDs := make([]driver.Value, len(batch.Payload.Messages))
-			for i, msg := range batch.Payload.Messages {
+			msgIDs := make([]driver.Value, len(state.Payload.Messages))
+			for i, msg := range state.Payload.Messages {
 				msgIDs[i] = msg.Header.ID
+				// We don't want to have to read the DB again if we want to query for the batch ID, or pins,
+				// so ensure the copy in our cache gets updated.
+				bp.data.UpdateMessageIfCached(ctx, msg)
```

Comment:
Want to make sure it's OK that we update the batchID in the cache before we write the batchID to the database itself (ie in case we somehow fail to update the database)... I think it's OK and we would come back around and try to write the same batchID on the second try. Just wanted to put a comment because I was staring at this for a bit.

Reply:
Yep, the only way to avoid this would be to push all of the cache updating (in all cases) to post-commit actions.
I convinced myself that wasn't required, but happy to discuss more.